instrumentation/grape/lib/opentelemetry/instrumentation/grape/event_handler.rb

```diff
@@ -0,0 +1,141 @@
+              handle_error(span, payload[:exception_object])
+            else
+              span.set_attribute(OpenTelemetry::SemanticConventions::Trace::HTTP_STATUS_CODE, payload[:endpoint].status)
+            end
+
+            span.finish
+            OpenTelemetry::Context.detach(token)
+          end
+
+          # Handles the endpoint_render.grape event
+          def endpoint_render(_name, start, _finish, _id, payload)
+            name = span_name(payload[:endpoint])
+            attributes = {
+              'component' => 'template',
+              'operation' => 'endpoint_render'
+            }
+            tracer.in_span(name, attributes: attributes, start_timestamp: start, kind: :server) do |span|
+              handle_error(span, payload[:exception_object]) if payload[:exception_object]
+            end
+          end
+
+          # Handles the endpoint_run_filters.grape events
+          def endpoint_run_filters(_name, start, finish, _id, payload)
+            filters = payload[:filters]
+            type = payload[:type]
+
+            # Prevent submitting empty filters
+            return if (!filters || filters.empty?) || !type || (finish - start).zero?
+
+            name = span_name(payload[:endpoint])
+            attributes = {
+              'component' => 'web',
+              'operation' => 'endpoint_run_filters',
+              'grape.filter.type' => type.to_s
+            }
+            tracer.in_span(name, attributes: attributes, start_timestamp: start, kind: :server) do |span|
+              handle_error(span, payload[:exception_object]) if payload[:exception_object]
+            end
+          end
+
+          # Handles the format_response.grape event
+          def format_response(_name, start, _finish, _id, payload)
+            endpoint = payload[:env]['api.endpoint']
+            name = span_name(endpoint)
+            attributes = {
+              'component' => 'template',
+              'operation' => 'format_response',
+              'grape.formatter.type' => formatter_type(payload[:formatter])
+            }
+            tracer.in_span(name, attributes: attributes, start_timestamp: start, kind: :server) do |span|
+              handle_error(span, payload[:exception_object]) if payload[:exception_object]
+            end
+          end
+
+          private
+
+          def tracer
+            Grape::Instrumentation.instance.tracer
+          end
+
+          def span_name(endpoint)
+            "#{api_instance(endpoint)} #{request_method(endpoint)} #{path(endpoint)}"
+          end
+
+          def run_attributes(payload)
+            endpoint = payload[:endpoint]
+            path = path(endpoint)
+            {
+              'component' => 'web',
+              'operation' => 'endpoint_run',
+              'grape.route.endpoint' => api_instance(endpoint),
+              'grape.route.path' => path,
+              'grape.route.method' => endpoint.options[:method].first,
```

Comment:
do we need this if we also set `HTTP_METHOD` ?

Reply:
I added http attributes because they were present in the [DDog instrumentation](https://github.com/DataDog/dd-trace-rb/blob/a02ccfc38123c087451f204ac6e822a6240216d5/lib/datadog/tracing/contrib/grape/endpoint.rb#L99-L105), so I was just porting that.

I'm all for removing repeated attributes (even more, I think I would remove the HTTP ones since they should be present in the HTTP instrumentation), but I would keep it like this for now, to add this comment in the upstream PR and ask the maintainers what direction they would like to take, wdyt?

Reply:
sounds good 👍 , personally I would keep the http attributes because of their standardized values, but it makes sense.